### ISSUE TITLE
The default parameters in the decorator are not detected properly

### ISSUE THREAD

**Issue by GaryLiuTelus** [microsoft/pyright#9209]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**
If a decorator return a function having default argument, and called without providing that argument, it will give an error. Looks the function signature did not take the default parameters. 

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.
![Image](https://github.com/user-attachments/assets/3c181f12-34ec-4ec3-8634-53f4fbfeda29)

```python
from typing import Callable, Optional
from functools import wraps

def dec(a: int) -> Callable[[Callable[[int], int]], Callable[[int, Optional[int]], int]]:
    def func_wrapper(func: Callable[[int], int]) -> Callable[[int, Optional[int]], int]:
        @wraps(func)
        def decorated(b: int, c: Optional[int] = None) -> int:
            result = func(b)
            if c:
                return result + a + c
            return result + a
        return decorated

    return func_wrapper

@dec(5)
def my_func(my_num: int) -> int:
    return my_num * 2


print(my_func(3)) # this give error of missing 1 positional argument
print(my_func(3, 5))
```
**The code execution results are the same as expected though.**

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?
vscode 1.94.0, with pylance 2024-10-07

![Image](https://github.com/user-attachments/assets/dfc48e9f-df4e-4bb1-aeff-7ee49ce28f81)

**Comment by erictraut** [comment]
Pyright is working correctly here.

Your decorator is annotated to return a type of `Callable[[int, Optional[int]], int]` which is equivalent to `Callable[[int, int | None], int]`. This annotation means that it's a callable that accepts two position-only parameters. The first must be assignable to `int`, and the second must be assignable to `int | None`.

I'm guessing that you are misinterpreting the meaning of `Optional`. It's a poorly-named symbol because it doesn't actually mean "optional". It means "a union with None". The `Optional` type form is no longer favored, and newer Python code typically uses `int | None` rather than `Optional[int]` for this reason. `int | None` also involves fewer characters, is easier to read, and doesn't require importing the `Optional` symbol from the `typing` module.

If your intent is to specify a callable type that has a default argument, you cannot do that with the `Callable` type form. You would need to use the more flexible "callback protocol" technique described [here](https://typing.readthedocs.io/en/latest/spec/callables.html#callback-protocols).

Here's what this would look like.

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgDRQAK4MYAxmCQFCiRTACuKViw4BnTNlz4A7iCIJRnTq3KjxAdSQwAFgBFqwIvxIwAFEzAt2JAJQAuTlCdQAJgagB9D61Ikvp0WoSYHoIOA8Ufgg7TBQYejAEGBjUfAAfKAA5MBRqKABeLJy6KAB6GygAWgA%2BWOSoADompTdgV2pWUyIUuIqawl9KGgBtYeIyIepR1IBdelm5qE0dfUNjGBmHZ3a2gSEPWXkEahBTPdYY8fIqKeHZ%2BbiZvtrlvQMjEy3t5wABQ4UzoJWDZHN8nK12uw5DBqC5TBQevEoBc6lAMtlcgUirlnnUvmDtiBqKJ1ljzvCQQTvkg2hdQVSwUSYPwQCgoESSSYoABqKBEHnI%2BkMpkstkc0m8ohKKki1mQ3BEGEuaXOWVs84HOQIY4gJQ-NydACsIIhYQ85LNkWidVxqXx7OozLllqiUAAVFAAExKBAgVKmM3kgDMNhBvv9gaBpiD9GNIKAA)

```python
from typing import Callable, Protocol
from functools import wraps

class WithDefault(Protocol):
    def __call__(self, my_num: int, opt: int | None = None, /) -> int: ...

def dec(a: int) -> Callable[[Callable[[int], int]], WithDefault]:
    def func_wrapper(func: Callable[[int], int]) -> WithDefault:
        @wraps(func)
        def decorated(b: int, c: int | None = None) -> int:
            result = func(b)
            if c:
                return result + a + c
            return result + a

        return decorated

    return func_wrapper

@dec(5)
def my_func(my_num: int) -> int:
    return my_num * 2

print(my_func(3))
print(my_func(3, 5))
```
